v3_async_wip/v3_async_wip/mqtt_client.py

```diff
@@ -438,28 +615,191 @@ async def wait_for_disconnect():
                 # Paho Disconnect
                 # NOTE: Paho disconnect shouldn't raise any exceptions
                 logger.debug("Attempting disconnect")
-                rc = await loop.run_in_executor(None, self._mqtt_client.disconnect)
-                message = mqtt.error_string(rc)
-                logger.debug("Disconnect returned rc {} - {}".format(rc, message))
+                rc = await self._loop.run_in_executor(None, self._mqtt_client.disconnect)
+                rc_msg = mqtt.error_string(rc)
+                logger.debug("Disconnect returned rc {} - {}".format(rc, rc_msg))
 
                 if rc == mqtt.MQTT_ERR_SUCCESS:
                     # Wait for disconnection to complete
-                    logger.debug("Waiting for disconnect response...")
-                    await disconnect_done
-                    logger.debug("Stopping Paho network loop")
-                    self._mqtt_client.loop_stop()
+                    logger.debug("Waiting for disconnect to complete...")
+                    async with self.disconnected_cond:
+                        await self.disconnected_cond.wait_for(lambda: not self.is_connected())
+                    logger.debug("Waiting for network loop to exit and clearing task")
+                    await self._network_loop
+                    self._network_loop = None
+                    # Wait slightly for tasks started by the on_disconnect handler to finish.
+                    # This will prevent warnings.
+                    # TODO: improve efficiency by being able to wait on something specific
+                    await asyncio.sleep(0.02)
                 elif rc == mqtt.MQTT_ERR_NO_CONN:
                     # This happens when we disconnect while already disconnected.
                     # In this implementation, it should only happen if Paho's inner state
                     # indicates we would like to be connected, but we actually aren't.
                     # We still want to do this disconnect however, because doing so changes
                     # Paho's state to indicate we no longer wish to be connected.
                     logger.debug("Early disconnect return (Already disconnected)")
-                    disconnect_done.cancel()
+                    if self._network_loop_running():
+                        # This block should never execute
+                        logger.warning("Network loop unexpectedly still running. Waiting")
+                        await self._network_loop
```

Comment:
This feels like you're taking something that _shouldn't_ happen and you're turning it into a hang. If the network loop is running, and we don't think it should be running, why do we think it's going to stop now?

Reply:
I agree. There's defensive coding, and then there's inventing problems to be worried about. I'll remove this